interceptors/tracing/interceptors_test.go

```diff
@@ -0,0 +1,317 @@
+	traceIDHeaderKey       = "traceid"
+	spanIDHeaderKey        = "spanid"
+)
+
+func extractFromContext(ctx context.Context, kind tracing.SpanKind) *mockSpan {
+	var m metadata.MD
+	if kind == tracing.SpanKindClient {
+		m, _ = metadata.FromOutgoingContext(ctx)
+	} else {
+		m, _ = metadata.FromIncomingContext(ctx)
+	}
+
+	traceIDValues := m.Get(traceIDHeaderKey)
+	if len(traceIDValues) == 0 {
+		return nil
+	}
+	spanIDValues := m.Get(spanIDHeaderKey)
+	if len(spanIDValues) == 0 {
+		return nil
+	}
+
+	return &mockSpan{
+		traceID: traceIDValues[0],
+		spanID:  spanIDValues[0],
+	}
+}
+
+func injectWithContext(ctx context.Context, span *mockSpan, kind tracing.SpanKind) context.Context {
+	var m metadata.MD
+	if kind == tracing.SpanKindClient {
+		m, _ = metadata.FromOutgoingContext(ctx)
+	} else {
+		m, _ = metadata.FromIncomingContext(ctx)
+	}
+	m = m.Copy()
+
+	m.Set(traceIDHeaderKey, span.traceID)
+	m.Set(spanIDHeaderKey, span.spanID)
+
+	ctx = metadata.NewOutgoingContext(ctx, m)
+	return ctx
+}
+
+func genID() string {
+	return strconv.FormatInt(atomic.AddInt64(&id, 1), 10)
+}
+
+// Implements Tracker
+type mockTracer struct {
+	spanStore map[string]*mockSpan
+}
+
+func (t *mockTracer) ListSpan(kind tracing.SpanKind) []*mockSpan {
+	var spans []*mockSpan
+	for _, v := range t.spanStore {
+		if v.kind == kind {
+			spans = append(spans, v)
+		}
+	}
+	return spans
+}
+
+func (t *mockTracer) Reset() {
+	t.spanStore = make(map[string]*mockSpan)
+}
+
+func newMockTracer() *mockTracer {
+	return &mockTracer{
+		spanStore: make(map[string]*mockSpan),
+	}
+}
+
+func (t *mockTracer) Start(ctx context.Context, spanName string, kind tracing.SpanKind) (context.Context, tracing.Span) {
+	span := mockSpan{
+		spanID:     genID(),
+		name:       spanName,
+		kind:       kind,
+		statusCode: codes.OK,
+	}
+
+	// parentSpan := spanFromContext(ctx)
+	parentSpan := extractFromContext(ctx, kind)
+	if parentSpan != nil {
+		// Fetch span from context as parent span
+		span.traceID = parentSpan.traceID
+		span.parentSpanID = parentSpan.spanID
+	} else {
+		span.traceID = genID()
+	}
+
+	t.spanStore[span.spanID] = &span
+
+	// ctx = contextWithSpan(ctx, &span)
+	if kind == tracing.SpanKindClient {
+		ctx = injectWithContext(ctx, &span, kind)
+	}
+	return ctx, &span
+}
+
+// Implements Span
+type mockSpan struct {
+	traceID      string
+	spanID       string
+	parentSpanID string
+
+	name string
+	kind tracing.SpanKind
+	end  bool
+
+	statusCode    codes.Code
+	statusMessage string
+
+	msgSendCounter     int
+	msgReceivedCounter int
+	eventNameList      []string
+	attributesList     [][]kv.KeyValue
+}
+
+func (s *mockSpan) SetAttributes(attrs ...kv.KeyValue) {
+	s.attributesList = append(s.attributesList, attrs)
+}
+
+func (s *mockSpan) End() {
+	s.end = true
+}
+
+func (s *mockSpan) SetStatus(code codes.Code, message string) {
+	s.statusCode = code
+	s.statusMessage = message
+}
+
+func (s *mockSpan) AddEvent(name string, attrs ...kv.KeyValue) {
```

Comment:
```suggestion
func (s *mockSpan) AddEvent(name string, attrs ...any) {
```
How bad would be to operate on interfaces (any) and expert implementation to deal with types?

NOTE: I know we could use generics, but in practice they have still some overhead, plus not everyone moved to Go 1.18, so we can add generic type of implementation later on (maybe even in non braking compatibility manner)

Reply:
If we would move to interfaces, we won't be needing `kv` package - Otel implementation can then use `"go.opentelemetry.io/otel/attribute"`